lms/djangoapps/courseware/tests/test_views.py

```diff
@@ -269,8 +269,8 @@ class IndexQueryTestCase(ModuleStoreTestCase):
     NUM_PROBLEMS = 20
 
     @ddt.data(
-        (ModuleStoreEnum.Type.mongo, 10, 172),
-        (ModuleStoreEnum.Type.split, 4, 168),
+        (ModuleStoreEnum.Type.mongo, 10, 171),
+        (ModuleStoreEnum.Type.split, 4, 167),
```

Comment:
Question: Why did this happen? Not complaining, but I don't understand.

Reply:
Since the number was going down I didn't really look into it. Will do and get back.

Reply:
Thank you. I'm mostly paranoid that there's some sequence-logic check that's no longer happening.

Reply:
I have added two commits which list the before and after queries being made in these tests. If you click the second one you will see the changes to the queries after the conversion.

1. One difference is that the code https://github.com/edx/edx-platform/blob/62ed654b31faed7f585562380972cbea80db8f99/lms/djangoapps/courseware/module_render.py#L759 is going into the `if` instead of the `else`. I do not think it should matter because these 3 XBlocks do not use the anonymous user id. Though I can set `requires_per_student_anonymous_id = True` just to be on the safe side. However, that does not change the count of queries.
2. The query that I think is missing after the conversion is https://github.com/edx/edx-platform/pull/25965/commits/1d756cb8b6a81a7622292fd2edf1e132edd84f81#diff-f3b98e9fb37256e372d064b8a6efd3ab578ef5107b70833fb495ecce36219d73L337. Though I am still trying to figure out where this is getting executed from.

Reply:
@ormsbee 

It turns out this is because of a caching bug that is no longer being hit inadvertently (pun intended). 😄 

There are 6 calls to `correct_modes_for_fbe()` when rendering this view: https://github.com/edx/edx-platform/blob/master/openedx/features/content_type_gating/helpers.py#L38

There is a `request_cached()` decorator on `modes_for_course()`: https://github.com/edx/edx-platform/blob/master/common/djangoapps/course_modes/models.py#L338

The `__repr__()` of `XModuleDescriptor()` also includes the whole runtime representation: https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1276-L1283

The first time it is called there is no `ModuleSystem` set on it so the cache_key becomes:
```
"common.djangoapps.course_modes.models.modes_for_course.<class 'common.djangoapps.course_modes.models.CourseMode'>.course.CourseDescriptorWithMixins(CombinedSystem(None,
CachingDescriptorSystem(<xmodule.modulestore.mongo.draft.DraftModuleStore object at 0x7fc91ec77370>, 'org.0/course_0/Run_0', ['i4x://org.0/course_0/course/Run_0',
'i4x://org.0/course_0/chapter/chapter_1', 'i4x://org.0/course_0/sequential/sequential_2'], <class 'xmodule.raw_module.RawDescriptor'>, ['i4x://org.0/course_0/chapter/chapter_1',
'i4x://org.0/course_0/sequential/sequential_2', 'i4x://org.0/course_0/vertical/vertical_3', 'i4x://org.0/course_0/problem/problem_4', 'i4x://org.0/course_0/problem/problem_5',
'i4x://org.0/course_0/problem/problem_6', 'i4x://org.0/course_0/problem/problem_7', 'i4x://org.0/course_0/problem/problem_8', 'i4x://org.0/course_0/problem/problem_9',
'i4x://org.0/course_0/problem/problem_10', 'i4x://org.0/course_0/problem/problem_11', 'i4x://org.0/course_0/problem/problem_12', 'i4x://org.0/course_0/problem/problem_13',
'i4x://org.0/course_0/problem/problem_14', 'i4x://org.0/course_0/problem/problem_15', 'i4x://org.0/course_0/problem/problem_16', 'i4x://org.0/course_0/problem/problem_17',
'i4x://org.0/course_0/problem/problem_18', 'i4x://org.0/course_0/problem/problem_19', 'i4x://org.0/course_0/problem/problem_20', 'i4x://org.0/course_0/problem/problem_21',
'i4x://org.0/course_0/problem/problem_22', 'i4x://org.0/course_0/problem/problem_23'])), KvsFieldData(MongoKeyValueStore({'wiki_slug': 'course_0'}, None,
[BlockUsageLocator(CourseLocator('org.0', 'course_0', 'Run_0', None, None), 'chapter', 'chapter_1')], {'display_name': 'Run 0', 'discussion_topics': {'General': {'id':
'i4x-org_0-course_0-course-Run_0'}}, 'tabs': [{'type': 'course_info', 'name': 'Home', 'course_staff_only': False}, {'type': 'courseware', 'name': 'Course', 'course_staff_only': False,
'link_func': <function CoursewareTab.__init__.<locals>.link_func at 0x7fc9074b5430>}, {'type': 'textbooks', 'name': 'Textbooks', 'course_staff_only': False}, {'type': 'discussion',
'name': 'Discussion', 'course_staff_only': False}, {'type': 'wiki', 'name': 'Wiki', 'course_staff_only': False}, {'type': 'progress', 'name': 'Progress', 'course_staff_only':
False}]})<{}, {}>), ScopeIds(user_id=None, block_type='course', def_id=BlockUsageLocator(CourseLocator('org.0', 'course_0', 'Run_0', None, None), 'course', 'Run_0'),
usage_id=BlockUsageLocator(CourseLocator('org.0', 'course_0', 'Run_0', None, None), 'course', 'Run_0'))).include_expired.True.only_selectable.False"
```

The second time it is called in `get_module_for_descriptor_internal()` and by that time the `ModuleSystem` has been set on it so the key changes to:
```
"common.djangoapps.course_modes.models.modes_for_course.<class
'common.djangoapps.course_modes.models.CourseMode'>.course.CourseDescriptorWithMixins(CombinedSystem(LmsModuleSystem{'request_token': 'f769d748784b11eba4a20242ac12000b', 'id_reader':
<xmodule.modulestore.xml.CourseLocationManager object at 0x7fc91dccfeb0>, '_services': {'fs': File system object, 'field-data':
LmsFieldData(ReadOnlyFieldData(<edx_when.field_data.DateLookupFieldData object at 0x7fc91f0a38b0>), KvsFieldData(<lms.djangoapps.courseware.model_data.DjangoKeyValueStore object at
0x7fc9070e7670>)), 'user': <common.djangoapps.xblock_django.user_service.DjangoXBlockUserService object at 0x7fc9070e7eb0>, 'verification':
<lms.djangoapps.verify_student.services.XBlockVerificationService object at 0x7fc91e395dc0>, 'proctoring': <edx_proctoring.services.ProctoringService object at 0x7fc91f02d2b0>,
'milestones': <milestones.services.MilestonesService object at 0x7fc9071567c0>, 'credit': <openedx.core.djangoapps.credit.services.CreditService object at 0x7fc907156490>,
'bookmarks': <openedx.core.djangoapps.bookmarks.services.BookmarksService object at 0x7fc907156bb0>, 'gating': <openedx.core.lib.gating.services.GatingService object at
0x7fc9071d1280>, 'grade_utils': <lms.djangoapps.grades.util_services.GradesUtilService object at 0x7fc9071d18e0>, 'user_state': <lms.djangoapps.courseware.services.UserStateService
object at 0x7fc9071d1850>, 'content_type_gating': <openedx.features.content_type_gating.services.ContentTypeGatingService object at 0x7fc9071d1d30>, 'completion':
<completion.services.CompletionService object at 0x7fc907188a00>, 'i18n': <class 'xmodule.modulestore.django.ModuleI18nService'>, 'library_tools':
<xmodule.library_tools.LibraryToolsService object at 0x7fc9070a0190>, 'partitions': <xmodule.partitions.partitions_service.PartitionService object at 0x7fc907188400>, 'settings':
<xmodule.services.SettingsService object at 0x7fc907110220>, 'user_tags': <lms.djangoapps.lms_xblock.runtime.UserTagsService object at 0x7fc907110100>, 'teams':
<lms.djangoapps.teams.services.TeamsService object at 0x7fc907110fa0>, 'teams_configuration': <xmodule.services.TeamsConfigurationService object at 0x7fc907110a90>, 'call_to_action':
<openedx.core.lib.xblock_services.call_to_action.CallToActionService object at 0x7fc907110f40>}, '_deprecated_per_instance_field_data': None, 'default_class': None, 'select': None,
'user_id': 1, 'mixologist': <xblock.runtime.Mixologist object at 0x7fc907110e50>, 'id_generator': <xmodule.modulestore.xml.CourseLocationManager object at 0x7fc91dccfeb0>, 'wrappers':
[functools.partial(<function wrap_xblock at 0x7fc925b3fee0>, 'LmsRuntime', extra_data={'course-id': 'org.0/course_0/Run_0'}, usage_id_serializer=<function
get_module_system_for_user.<locals>.<lambda> at 0x7fc9070d1b80>, request_token='f769d748784b11eba4a20242ac12000b'), functools.partial(<function replace_static_urls at 0x7fc925b4d8b0>,
None, course_id=CourseLocator('org.0', 'course_0', 'Run_0', None, None), static_asset_path=''), functools.partial(<function replace_course_urls at 0x7fc925b4d820>,
CourseLocator('org.0', 'course_0', 'Run_0', None, None)), functools.partial(<function replace_jump_to_id_urls at 0x7fc925b4d040>, CourseLocator('org.0', 'course_0', 'Run_0', None,
None), '/courses/org.0/course_0/Run_0/jump_to_id/'), functools.partial(<function display_access_messages at 0x7fc9254e9040>, <User: robot0>), functools.partial(<function
course_expiration_wrapper at 0x7fc926004670>, <User: robot0>), functools.partial(<function offer_banner_wrapper at 0x7fc925517a60>, <User: robot0>)], 'wrappers_asides': [],
'STATIC_URL': '/static/', 'xqueue': {'interface': <capa.xqueue_interface.XQueueInterface object at 0x7fc925d93400>, 'construct_callback': <function
get_module_system_for_user.<locals>.make_xqueue_callback at 0x7fc9070d1ee0>, 'default_queuename': 'org.0-course_0', 'waittime': 5}, 'track_function': <function
make_track_function.<locals>.function at 0x7fc9070d1040>, 'filestore': OSFS('/tmp/tmpp_frzs02/course_0'), 'get_module': <function get_module_system_for_user.<locals>.inner_get_module
at 0x7fc9070d1ca0>, 'render_template': <function render_to_string at 0x7fc927e94310>, 'DEBUG': False, 'debug': False, 'HOSTNAME': 'edx.org', 'hostname': 'edx.org', 'seed': 1,
'replace_urls': functools.partial(<function replace_static_urls at 0x7fc93a91e700>, data_directory=None, course_id=CourseLocator('org.0', 'course_0', 'Run_0', None, None),
static_asset_path=''), 'node_path': '/edx/app/edxapp/edx-platform/common/static/js/vendor:/edx/app/edxapp/edx-platform/node_modules', 'anonymous_student_id':
'de619ab51c7f4e9c7216b4644c24f3b5', 'course_id': CourseLocator('org.0', 'course_0', 'Run_0', None, None), 'user_is_staff': False, 'publish': <function
get_module_system_for_user.<locals>.publish at 0x7fc9070d1af0>, 'cache': <django.core.cache.DefaultCacheProxy object at 0x7fc93e117400>, 'can_execute_unsafe_code': <function
get_module_system_for_user.<locals>.<lambda> at 0x7fc91df19790>, 'get_python_lib_zip': <function get_module_system_for_user.<locals>.<lambda> at 0x7fc91df19ca0>,
'replace_course_urls': functools.partial(<function replace_course_urls at 0x7fc93a91e550>, course_key=CourseLocator('org.0', 'course_0', 'Run_0', None, None)),
'replace_jump_to_id_urls': functools.partial(<function replace_jump_to_id_urls at 0x7fc93a91e4c0>, course_id=CourseLocator('org.0', 'course_0', 'Run_0', None, None),
jump_to_id_base_url='/courses/org.0/course_0/Run_0/jump_to_id/'), 'error_descriptor_class': <class 'xmodule.error_module.NonStaffErrorBlock'>, 'xmodule_instance': None,
'get_real_user': <function user_by_anonymous_id at 0x7fc9275d9160>, 'user_location': None, 'get_user_role': <function get_module_system_for_user.<locals>.<lambda> at 0x7fc91df19af0>,
'descriptor_runtime': CachingDescriptorSystem(<xmodule.modulestore.mongo.draft.DraftModuleStore object at 0x7fc91ec77370>, 'org.0/course_0/Run_0',
['i4x://org.0/course_0/course/Run_0', 'i4x://org.0/course_0/chapter/chapter_1', 'i4x://org.0/course_0/sequential/sequential_2'], <class 'xmodule.raw_module.RawDescriptor'>,
['i4x://org.0/course_0/chapter/chapter_1', 'i4x://org.0/course_0/sequential/sequential_2', 'i4x://org.0/course_0/vertical/vertical_3', 'i4x://org.0/course_0/problem/problem_4',
'i4x://org.0/course_0/problem/problem_5', 'i4x://org.0/course_0/problem/problem_6', 'i4x://org.0/course_0/problem/problem_7', 'i4x://org.0/course_0/problem/problem_8',
'i4x://org.0/course_0/problem/problem_9', 'i4x://org.0/course_0/problem/problem_10', 'i4x://org.0/course_0/problem/problem_11', 'i4x://org.0/course_0/problem/problem_12',
'i4x://org.0/course_0/problem/problem_13', 'i4x://org.0/course_0/problem/problem_14', 'i4x://org.0/course_0/problem/problem_15', 'i4x://org.0/course_0/problem/problem_16',
'i4x://org.0/course_0/problem/problem_17', 'i4x://org.0/course_0/problem/problem_18', 'i4x://org.0/course_0/problem/problem_19', 'i4x://org.0/course_0/problem/problem_20',
'i4x://org.0/course_0/problem/problem_21', 'i4x://org.0/course_0/problem/problem_22', 'i4x://org.0/course_0/problem/problem_23']), 'rebind_noauth_module_to_user': <function
get_module_system_for_user.<locals>.rebind_noauth_module_to_user at 0x7fc9070d18b0>, 'position': None, 'user_is_admin': False, 'user_is_beta_tester': False, 'days_early_for_beta':
None}, CachingDescriptorSystem(<xmodule.modulestore.mongo.draft.DraftModuleStore object at 0x7fc91ec77370>, 'org.0/course_0/Run_0', ['i4x://org.0/course_0/course/Run_0',
'i4x://org.0/course_0/chapter/chapter_1', 'i4x://org.0/course_0/sequential/sequential_2'], <class 'xmodule.raw_module.RawDescriptor'>, ['i4x://org.0/course_0/chapter/chapter_1',
'i4x://org.0/course_0/sequential/sequential_2', 'i4x://org.0/course_0/vertical/vertical_3', 'i4x://org.0/course_0/problem/problem_4', 'i4x://org.0/course_0/problem/problem_5',
'i4x://org.0/course_0/problem/problem_6', 'i4x://org.0/course_0/problem/problem_7', 'i4x://org.0/course_0/problem/problem_8', 'i4x://org.0/course_0/problem/problem_9',
'i4x://org.0/course_0/problem/problem_10', 'i4x://org.0/course_0/problem/problem_11', 'i4x://org.0/course_0/problem/problem_12', 'i4x://org.0/course_0/problem/problem_13',
'i4x://org.0/course_0/problem/problem_14', 'i4x://org.0/course_0/problem/problem_15', 'i4x://org.0/course_0/problem/problem_16', 'i4x://org.0/course_0/problem/problem_17',
'i4x://org.0/course_0/problem/problem_18', 'i4x://org.0/course_0/problem/problem_19', 'i4x://org.0/course_0/problem/problem_20', 'i4x://org.0/course_0/problem/problem_21',
'i4x://org.0/course_0/problem/problem_22', 'i4x://org.0/course_0/problem/problem_23'])), LmsFieldData(ReadOnlyFieldData(<edx_when.field_data.DateLookupFieldData object at
0x7fc907110ee0>), LmsFieldData(ReadOnlyFieldData(<edx_when.field_data.DateLookupFieldData object at 0x7fc91f0a38b0>),
KvsFieldData(<lms.djangoapps.courseware.model_data.DjangoKeyValueStore object at 0x7fc9070e7670>))), ScopeIds(user_id=1, block_type='course',
def_id=BlockUsageLocator(CourseLocator('org.0', 'course_0', 'Run_0', None, None), 'course', 'Run_0'), usage_id=BlockUsageLocator(CourseLocator('org.0', 'course_0', 'Run_0', None,
None), 'course', 'Run_0'))).include_expired.True.only_selectable.False"
```

However, after `XModuleDescriptor` is no longer a superclass of `CourseBlock`, its representation does not include the runtime representation:
```
"common.djangoapps.course_modes.models.modes_for_course.<class 'common.djangoapps.course_modes.models.CourseMode'>.course.<CourseBlockWithMixins @7C45 license=None, parent=None,
name=None, tags=[], display_name='Run 0', course_edit_method='Studio', days_early_for_beta=None, due=None, edxnotes=False, edxnotes_visibility=True, giturl=None, graceperiod=None,
graded=False, group_access={}, in_entrance_exam=False, matlab_api_key=None, max_attempts=None, rerandomize='never', self_paced=False, show_correctness='always',
show_reset_button=False, showanswer='finished', start=datetime.datetime(2030, 1, 1, 0, 0, tzinfo=<UTC>), static_asset_path='', use_latex_compiler=False, user_partitions=[],
video_auto_advance=False, video_bumper={}, video_speed_optimizations=True, visible_to_staff_only=False, xqa_key=None, chrome=None, default_tab=None, format=None, hide_from_toc=False,
source_file=None, children=[BlockUsageLocator(CourseLocator('org.0', 'course_0', 'Run_0', None, None), 'chapter', 'chapter_1')], default_time_limit_minutes=None, exam_review_rules='',
hide_after_due=False, is_entrance_exam=False, is_onboarding_exam=False, is_practice_exam=False, is_proctored_enabled=False, is_time_limited=False, position=None, xml_attributes={},
advanced_modules=[], advertised_start=None, allow_anonymous=True, allow_anonymous_to_peers=False, allow_proctoring_opt_out=False, allow_public_wiki_access=False,
allow_unsupported_xblocks=False, announcement=None, banner_image='images_course_image.jpg', catalog_visibility='both', ccx_connector='', cert_html_view_enabled=True,
cert_html_view_overrides={}, cert_name_long='', cert_name_short='', certificate_available_date=None, certificates={}, certificates_display_behavior='end',
certificates_show_before_end=False, cohort_config={}, cosmetic_display_price=0, course_image='images_course_image.jpg', course_survey_name=None, course_survey_required=False,
course_visibility='private', create_zendesk_tickets=True, css_class='', disable_progress_graph=False, discussion_blackouts=[], discussion_link=None, discussion_sort_alpha=False,
discussion_topics={'General': {'id': 'i4x-org_0-course_0-course-Run_0'}}, display_coursenumber='', display_organization=None, due_date_display_format=None, enable_ccx=False,
enable_proctored_exams=False, enable_subsection_gating=False, enable_timed_exams=False, end=None, end_of_course_survey_url=None, enrollment_domain=None, enrollment_end=None,
enrollment_start=None, entrance_exam_enabled=False, entrance_exam_id=None, entrance_exam_minimum_score_pct=65, grading_policy={'GRADER': [{'type': 'Homework', 'short_label': 'HW',
'min_count': 12, 'drop_count': 2, 'weight': 0.15}, {'type': 'Lab', 'min_count': 12, 'drop_count': 2, 'weight': 0.15}, {'type': 'Midterm Exam', 'short_label': 'Midterm', 'min_count':
1, 'drop_count': 0, 'weight': 0.3}, {'type': 'Final Exam', 'short_label': 'Final', 'min_count': 1, 'drop_count': 0, 'weight': 0.4}], 'GRADE_CUTOFFS': {'Pass': 0.5}},
hide_progress_tab=None, highlights_enabled_for_messaging=False, html_textbooks=[], info_sidebar_name='Course Handouts', instructor_info={'instructors': []}, invitation_only=False,
is_new=None, issue_badges=True, language=None, learning_info=[], lti_passports=[], max_student_enrollments_allowed=None, minimum_grade_credit=0.8, mobile_available=True,
no_grade=False, other_course_settings={}, pdf_textbooks=[], pre_requisite_courses=[], proctoring_escalation_email=None, proctoring_provider='null', remote_gradebook={},
show_calculator=False, show_timezone=True, social_sharing_url=None, tabs=[<lms.djangoapps.courseware.tabs.CourseInfoTab object at 0x7f2e4676f1c0>,
<lms.djangoapps.courseware.tabs.CoursewareTab object at 0x7f2e4676f220>, <lms.djangoapps.courseware.tabs.TextbookTabs object at 0x7f2e4676f250>,
<lms.djangoapps.discussion.plugins.DiscussionTab object at 0x7f2e4676f2e0>, <lms.djangoapps.course_wiki.tab.WikiTab object at 0x7f2e4676f310>,
<lms.djangoapps.courseware.tabs.ProgressTab object at 0x7f2e4676f370>], teams_configuration=<TeamsConfig default_max_team_size=50 teamsets=[]>, textbooks=[],
video_thumbnail_image='images_course_image.jpg', video_upload_pipeline={}, wiki_slug='course_0'>.include_expired.True.only_selectable.False"
```

So the second call to `modes_for_course()` in `get_module_for_descriptor_internal()` hits the cache.

Having said so it looks like the `CourseBlock` is constructed multiple times in this request which results in a different key and so a cache miss. There are 7 calls in this test to `modes_for_course()` and after the conversion only 2 hit the request cache.

Reply:
😮  Nice! Thanks for tracking that down!